tests/plotting/test_widgets.py

```diff
@@ -2,136 +2,169 @@
 import pytest
 import vtk
 
-import pyvista
+import pyvista as pv
 from pyvista import examples
+from pyvista.plotting.affine_widget import DARK_YELLOW, get_angle, ray_plane_intersection
 
 # skip all tests if unable to render
 pytestmark = pytest.mark.skip_plotting
 
 
+def r_mat_to_euler_angles(R):
+    """
+    Extract Euler angles from a 3x3 rotation matrix using the ZYX sequence.
+    Returns the angles in radians.
+    """
+
+    # Check for gimbal lock: singular cases
+    if abs(R[2, 0]) != 1:
+        # Not at the singularities
+        pitch = -np.arcsin(R[2, 0])
+        roll = np.arctan2(R[2, 1] / np.cos(pitch), R[2, 2] / np.cos(pitch))
+        yaw = np.arctan2(R[1, 0] / np.cos(pitch), R[0, 0] / np.cos(pitch))
+    else:
+        # Gimbal lock exists
+        yaw = 0  # Set yaw to 0 and compute the others
+        if R[2, 0] == -1:  # Directly looking up
+            pitch = np.pi / 2
+            roll = yaw + np.arctan2(R[0, 1], R[0, 2])
+        else:  # Directly looking down
+            pitch = -np.pi / 2
+            roll = -yaw + np.arctan2(-R[0, 1], -R[0, 2])
```

Comment:
The if statement is not a negative but an affirmative, which is easier for the reader to understand.
```suggestion
    # Check for gimbal lock: singular cases
    if abs(R[2, 0]) == 1:
        # Gimbal lock exists
        yaw = 0  # Set yaw to 0 and compute the others
        if R[2, 0] == -1:  # Directly looking up
            pitch = np.pi / 2
            roll = yaw + np.arctan2(R[0, 1], R[0, 2])
        else:  # Directly looking down
            pitch = -np.pi / 2
            roll = -yaw + np.arctan2(-R[0, 1], -R[0, 2])
    else:
        # Not at the singularities
        pitch = -np.arcsin(R[2, 0])
        roll = np.arctan2(R[2, 1] / np.cos(pitch), R[2, 2] / np.cos(pitch))
        yaw = np.arctan2(R[1, 0] / np.cos(pitch), R[0, 0] / np.cos(pitch))
```